File: src/retry_logic.c
/*
 * retry_logic.c - Common retry logic for API calls
 *
 * Centralizes retryable error determination and context length error handling
 * to avoid duplication across provider implementations.
 */

#include "retry_logic.h"
#include <string.h>
#include <stdlib.h>
#include <ctype.h>
#include <curl/curl.h>

/**
 * Determine if an HTTP error is retryable based on status code
 *
 * Retryable HTTP errors:
 * - 429: Rate limit (Too Many Requests)
 * - 408: Request timeout
 * - 5xx: Server errors
 *
 * @param http_status HTTP status code
 * @return 1 if retryable, 0 otherwise
 */
int is_http_error_retryable(long http_status) {
    return (http_status == 429 || http_status == 408 || http_status >= 500);
}


/**
 * Determine if a curl error is retryable
 *
 * Retryable curl errors are typically transient network issues:
 * - Connection issues (COULDNT_CONNECT, RECV_ERROR, SEND_ERROR, GOT_NOTHING)
 * - Timeouts (OPERATION_TIMEDOUT)
 * - SSL issues (SSL_CONNECT_ERROR)
 * - HTTP2/HTTP3 protocol layer issues (HTTP2, HTTP2_STREAM)
 *
 * @param curl_code curl error code
 * @return 1 if retryable, 0 otherwise
 */
int is_curl_error_retryable(CURLcode curl_code) {
    return (curl_code == CURLE_COULDNT_CONNECT ||
            curl_code == CURLE_OPERATION_TIMEDOUT ||
            curl_code == CURLE_RECV_ERROR ||
            curl_code == CURLE_SEND_ERROR ||
            curl_code == CURLE_SSL_CONNECT_ERROR ||
            curl_code == CURLE_GOT_NOTHING ||
            curl_code == CURLE_HTTP2 ||
            curl_code == CURLE_HTTP2_STREAM);
}

/**
 * Check if an error message indicates a context length error
 *
 * Detects various forms of context length/token limit errors from API providers.
 * Performs case-insensitive matching to handle variations in error messages.
 *
 * @param error_message Error message text (may be NULL)
 * @param error_type Error type from API response (may be NULL)
 * @return 1 if context length error, 0 otherwise
 */
int is_context_length_error(const char *error_message, const char *error_type) {
    if (!error_message) {
        return 0;
    }

    // Create lowercase copy of error message for case-insensitive matching
    size_t msg_len = strlen(error_message);
    char *msg_lower = malloc(msg_len + 1);
    if (msg_lower) {
        for (size_t i = 0; i < msg_len; i++) {
            msg_lower[i] = (char)tolower((unsigned char)error_message[i]);
        }
        msg_lower[msg_len] = '\0';
    }

    // Use lowercased message if available, otherwise fall back to original
    const char *msg_to_check = msg_lower ? msg_lower : error_message;
    int is_context_error = 0;

    // Check for context length/token limit patterns in error message (lowercased)
    if (strstr(msg_to_check, "maximum context length") != NULL ||
        (strstr(msg_to_check, "context length") != NULL && strstr(msg_to_check, "tokens") != NULL) ||
        strstr(msg_to_check, "too many tokens") != NULL ||
        // OpenAI token limit exceeded pattern
        strstr(msg_to_check, "exceeded model token limit") != NULL ||
        strstr(msg_to_check, "token limit") != NULL ||
        // LiteLLM / Bedrock patterns
        strstr(msg_to_check, "contextwindowexceedederror") != NULL ||
        strstr(msg_to_check, "context window error") != NULL ||
        strstr(msg_to_check, "input is too long") != NULL) {
        is_context_error = 1;
    }

    // Check for invalid_request_error with token-related message
    if (!is_context_error && error_type &&
        strcasecmp(error_type, "invalid_request_error") == 0 &&
        strstr(msg_to_check, "tokens") != NULL) {
        is_context_error = 1;
    }

    // msg_lower is either allocated buffer or NULL (never error_message)
    // free(NULL) is safe (no-op), but check explicitly for clarity
    if (msg_lower) {
        free(msg_lower);
    }
    return is_context_error;
}

/**
 * Get the standard context length error message
 *
 * Returns a user-friendly explanation of context length errors with suggestions.
 * Caller must free the returned string.
 *
 * @return Newly allocated string with context length error message
 */
char *get_context_length_error_message(void) {
    return strdup(
        "Context length exceeded. The conversation has grown too large for the model's memory. "
        "Try starting a new conversation or reduce the amount of code/files being discussed."
    );
}

File: zig/retry_logic.zig
//! retry_logic.zig — Retry/backoff logic for HTTP API calls
//!
//! Zig port of src/retry_logic.c.
//!
//! Provides:
//! - `shouldRetry(status_code)` — true for 408, 429, 5xx
//! - `calcDelay(attempt, config)` — exponential backoff with ±jitter
//! - `isContextLengthError(message, error_type)` — detect token-limit errors
//! - `contextLengthErrorMessage()` — user-facing explanation string

const std = @import("std");

// ---------------------------------------------------------------------------
// Configuration
// ---------------------------------------------------------------------------

pub const RetryConfig = struct {
    /// Maximum number of retry attempts (not counting the first try).
    max_retries: u32 = 3,
    /// Base delay for the first retry, in milliseconds.
    base_delay_ms: u64 = 1_000,
    /// Upper cap on the computed delay, in milliseconds.
    max_delay_ms: u64 = 30_000,
    /// Fraction of the computed delay that can be shaved off as jitter.
    /// 0.25 → up to 25% reduction (mirrors C implementation).
    jitter_factor: f64 = 0.25,
};

pub const default_config = RetryConfig{};

// ---------------------------------------------------------------------------
// Core predicates
// ---------------------------------------------------------------------------

/// Returns true for HTTP status codes that warrant a retry.
///
/// Retryable status codes (matching C implementation):
///   429  — Too Many Requests (rate limit)
///   408  — Request Timeout
///   5xx  — Server errors
pub fn shouldRetry(status_code: u32) bool {
    return status_code == 429 or status_code == 408 or status_code >= 500;
}

/// Calculate the delay (in milliseconds) before the `attempt`-th retry.
///
/// Uses exponential backoff:  base * 2^(attempt-1),  capped at max_delay_ms,
/// then reduces by a random jitter of up to `jitter_factor` fraction.
///
/// `attempt` is 1-based: pass 1 for the first retry, 2 for the second, etc.
pub fn calcDelay(attempt: u32, config: RetryConfig) u64 {
    if (attempt == 0) return 0;

    // 2^(attempt-1) — use saturating shift to avoid overflow for large attempts
    const shift: u6 = @intCast(@min(attempt - 1, 62));
    const multiplier: u64 = @as(u64, 1) << shift;

    // Saturating multiply avoids overflow when base_delay_ms * multiplier wraps
    const raw = std.math.mul(u64, config.base_delay_ms, multiplier) catch config.max_delay_ms;
    const capped: u64 = @min(raw, config.max_delay_ms);

    // Jitter: reduce by a random fraction in [0, jitter_factor)
    // We use a simple PRNG seeded from the system's random source.
    var rng = std.rand.DefaultPrng.init(@as(u64, @bitCast(std.time.milliTimestamp())));
    const random = rng.random();
    const jitter_max: f64 = @as(f64, @floatFromInt(capped)) * config.jitter_factor;
    const jitter: u64 = @intFromFloat(random.float(f64) * jitter_max);

    return if (capped > jitter) capped - jitter else 0;
}

// ---------------------------------------------------------------------------
// Context-length error detection
// ---------------------------------------------------------------------------

/// Returns true if the error message or type indicates a token/context-length
/// limit was exceeded.
///
/// Performs a case-insensitive search for well-known patterns from OpenAI,
/// Anthropic, AWS Bedrock and LiteLLM.
pub fn isContextLengthError(message: ?[]const u8, error_type: ?[]const u8) bool {
    const msg = message orelse return false;

    // Inline ASCII-to-lower helper (avoids heap allocation)
    const patterns = [_][]const u8{
        "maximum context length",
        "context length exceeded",
        "too many tokens",
        "exceeded model token limit",
        "token limit",
        "contextwindowexceedederror",
        "context window error",
        "input is too long",
    };

    // Build a lowercase copy on the stack for comparison (up to 4 KiB)
    var lower_buf: [4096]u8 = undefined;
    const copy_len = @min(msg.len, lower_buf.len);
    for (msg[0..copy_len], 0..) |ch, i| {
        lower_buf[i] = std.ascii.toLower(ch);
    }
    const lower = lower_buf[0..copy_len];

    for (patterns) |pat| {
        if (std.mem.indexOf(u8, lower, pat) != null) return true;
    }

    // Special: "context length" + "tokens" together anywhere in the message
    if (std.mem.indexOf(u8, lower, "context length") != null and
        std.mem.indexOf(u8, lower, "tokens") != null)
    {
        return true;
    }

    // invalid_request_error with token mention
    if (error_type) |et| {
        if (std.ascii.eqlIgnoreCase(et, "invalid_request_error") and
            std.mem.indexOf(u8, lower, "tokens") != null)
        {
            return true;
        }
    }

    return false;
}

/// Return the standard user-facing context length error message.
///
/// The caller owns the returned slice and must free it with `allocator.free`.
pub fn contextLengthErrorMessage(allocator: std.mem.Allocator) ![]u8 {
    return allocator.dupe(u8,
        "Context length exceeded. The conversation has grown too large for " ++
        "the model's memory. Try starting a new conversation or reduce the " ++
        "amount of code/files being discussed.");
}

// ---------------------------------------------------------------------------
// Tests
// ---------------------------------------------------------------------------

test "shouldRetry — retryable codes" {
    try std.testing.expect(shouldRetry(429));
    try std.testing.expect(shouldRetry(408));
    try std.testing.expect(shouldRetry(500));
    try std.testing.expect(shouldRetry(502));
    try std.testing.expect(shouldRetry(503));
    try std.testing.expect(shouldRetry(504));
}

test "shouldRetry — non-retryable codes" {
    try std.testing.expect(!shouldRetry(200));
    try std.testing.expect(!shouldRetry(400));
    try std.testing.expect(!shouldRetry(401));
    try std.testing.expect(!shouldRetry(403));
    try std.testing.expect(!shouldRetry(404));
}

test "calcDelay — first attempt baseline" {
    const cfg = RetryConfig{ .base_delay_ms = 1000, .max_delay_ms = 30_000, .jitter_factor = 0 };
    // With zero jitter the delay should equal the capped base * 2^(attempt-1)
    try std.testing.expectEqual(@as(u64, 1000), calcDelay(1, cfg));
    try std.testing.expectEqual(@as(u64, 2000), calcDelay(2, cfg));
    try std.testing.expectEqual(@as(u64, 4000), calcDelay(3, cfg));
}

test "calcDelay — cap at max_delay_ms" {
    const cfg = RetryConfig{ .base_delay_ms = 1000, .max_delay_ms = 5000, .jitter_factor = 0 };
    // 2^4 * 1000 = 16000 > 5000 → should be capped
    const d = calcDelay(5, cfg);
    try std.testing.expect(d <= 5000);
}

test "calcDelay — jitter reduces delay" {
    const cfg = RetryConfig{ .base_delay_ms = 1000, .max_delay_ms = 30_000, .jitter_factor = 0.25 };
    // With 25% jitter the result must be in [750, 1000]
    const d = calcDelay(1, cfg);
    try std.testing.expect(d <= 1000);
    try std.testing.expect(d >= 750);
}

test "calcDelay — attempt=0 returns 0" {
    try std.testing.expectEqual(@as(u64, 0), calcDelay(0, default_config));
}

test "isContextLengthError — null message" {
    try std.testing.expect(!isContextLengthError(null, null));
}

test "isContextLengthError — OpenAI pattern" {
    try std.testing.expect(isContextLengthError("maximum context length exceeded", null));
    try std.testing.expect(isContextLengthError("This exceeds the model token limit", null));
    try std.testing.expect(isContextLengthError("Too many tokens in request", null));
}

test "isContextLengthError — Anthropic pattern" {
    try std.testing.expect(isContextLengthError("context length is 200000 tokens but input is too long", null));
    try std.testing.expect(isContextLengthError("input is too long", null));
}

test "isContextLengthError — Bedrock / LiteLLM pattern" {
    try std.testing.expect(isContextLengthError("ContextWindowExceededError thrown", null));
    try std.testing.expect(isContextLengthError("context window error occurred", null));
}

test "isContextLengthError — invalid_request_error with tokens" {
    try std.testing.expect(isContextLengthError("Request has too many tokens", "invalid_request_error"));
}

test "isContextLengthError — unrelated error" {
    try std.testing.expect(!isContextLengthError("Internal server error", null));
    try std.testing.expect(!isContextLengthError("Unauthorized", "auth_error"));
}

test "isContextLengthError — case insensitive" {
    try std.testing.expect(isContextLengthError("MAXIMUM CONTEXT LENGTH exceeded", null));
    try std.testing.expect(isContextLengthError("Token Limit Reached", null));
}

test "contextLengthErrorMessage allocates" {
    const msg = try contextLengthErrorMessage(std.testing.allocator);
    defer std.testing.allocator.free(msg);
    try std.testing.expect(msg.len > 0);
    try std.testing.expect(std.mem.indexOf(u8, msg, "Context length exceeded") != null);
}
